scrapy_poet/injection.py

```diff
@@ -0,0 +1,323 @@
+            if not provided_classes:
+                continue
+
+            kwargs = andi.plan(
+                provider,
+                is_injectable=is_injectable,
+                externally_provided=scrapy_provided_dependencies,
+                full_final_kwargs=False,
+            ).final_kwargs(scrapy_provided_dependencies)
+            results = yield maybeDeferred_coro(provider, set(provided_classes),
+                                               **kwargs)
+            extra_classes = results.keys() - provided_classes
+            if extra_classes:
+                raise UndeclaredProvidedTypeError(
+                    f"{provider} has returned {extra_classes} but they're not "
+                    f"listed as provided classes {provider.provided_classes}"
+                )
+            instances.update(results)
+
+        raise returnValue(instances)
+
+    @inlineCallbacks
+    def build_callback_dependencies(self, request: Request, response: Response):
+        """
+        Scan the configured callback for this request looking for the
+        dependencies and build the corresponding instances. Return a kwargs
+        dictionary with the built instances.
+        """
+        plan = self.build_plan(request)
+        provider_instances = yield from self.build_instances(request, response, plan)
+        callback_kwargs = plan.final_kwargs(provider_instances)
+        raise returnValue(callback_kwargs)
+
+
+def check_all_providers_are_callable(providers):
+    for provider in providers:
+        if not callable(provider):
+            raise NonCallableProviderError(
+                f"The provider {type(provider)} is not callable. "
+                f"It must implement '__call__' method"
+            )
+
+
+def load_provider_classes(settings: Settings) -> List[PageObjectInputProvider]:
+    providers = settings.getlist('SCRAPY_POET_PROVIDER_CLASSES') or PROVIDERS
+    logger.info(f"Loading providers {providers}")
+    result = []
+    for cls in settings.getlist('SCRAPY_POET_PROVIDER_CLASSES') or PROVIDERS:
+        if not callable(cls):
+            cls = load_object(cls)
+
+        result.append(cls)
+
+    return result
+
+
+def is_class_provided_by_any_provider_fn(providers: List[PageObjectInputProvider]
+                                         ) -> Callable[[Callable], bool]:
+    """
+    Return a function of type ``Callable[[Type], bool]`` that return
+    True if the given type is provided by any of the registered providers.
+
+    The attribute ``provided_classes`` from each provided is used.
+    This attribute can be a ``set`` or a ``Callable``. All sets are
+    joined together for efficiency.
+    """
+    sets_of_types: Set[Callable] = set()  # caching all sets found
+    individual_is_callable: List[Callable[[Callable], bool]] = [sets_of_types.__contains__]
+    for provider in providers:
+        provided_classes = provider.provided_classes
+
+        if isinstance(provided_classes, Set):
```

Comment:
does it work with frozenset?

Reply:
I'm asking because it is quiite natural to define Provider.povided_classes as frozenset - it is not supposed to be mutable.

Reply:
Good point. Implemented in https://github.com/scrapinghub/scrapy-poet/pull/41/commits/d0432e3da5885de37c21131e763540bb879ff1e3